File: C++/problem0670.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <cmath>
#include <ctime>
#include <cstdlib>
#include <windows.h>
#include <string>
#include <numeric>

using namespace std;

class Solution {
public:
    int maximumSwap(int num) {
        vector<int> num1, num2;
        int temp = num;
        while (temp)
        {
            num1.push_back(temp%10);
            temp /= 10;
        }
        num2 = num1;
        reverse(num1.begin(), num1.end());
        // sort(num2.begin(), num2.end());
        // sort(num2.begin(), num2.end(), less<int>());
        // reverse(num2.begin(), num2.end());
        sort(num2.begin(), num2.end(), greater<int>());
        
        if (num1 == num2)
        {
            return num;
        }

        int k;
        for (k = 0; k < num1.size(); ++k)
        {
            if (num1[k] != num2[k])
            {
                break;
            }
        }
        // cout << k << endl;
        // cout << num2[k] << endl;

        int p;
        for (p = num1.size()-1; k > -1; --p)
        {
            // cout << num1[p] << num2[k] << endl;
            if (num1[p] == num2[k])
            {
                break;
            }
        }
        // cout << p << endl;
        // cout << num1[k] << '\t' << num1[p] << endl;
        // int tmp = num1[k];
        // num1[k] = num1[p];
        // num1[p] = tmp;
        num1[k] = num1[k] + num1[p];
        num1[p] = num1[k] - num1[p];
        num1[k] = num1[k] - num1[p];

        // vector<int>::iterator it;
        // for (it = num1.begin(); it != num1.end(); ++it)
        // {
        //     cout << *it << '\t';
        // }
        // cout << endl;

        int ans = 0;
        for (int k = 0; k < num1.size(); ++k)
        {
            // cout << num1[k] << '\t';
            ans *= 10;
            ans += num1[k];
        }
        return ans;

    }
};

int main()
{
    int num = 2736;
    // int num = 9973;
    Solution solu;
    int ans = solu.maximumSwap(num);
    cout << ans << endl;
    return 0;
}

File: Python/problem0670.py
# -*- coding: utf-8 -*-
"""
Created on Mon Jun 24 10:58:47 2019

@author: Administrator
"""

class Solution:
    def maximumSwap(self, num: int) -> int:
#        tmp = [int(x) for x in str(num)]
#        for k in range(len(tmp)):
#            if tmp[k] == max(tmp[k:]):
#                pass
#            else:
##                print(tmp)
##                print(k, max(tmp[k:]))
##                print(-1)
##                print(tmp[k])
##                print(tmp[tmp.index(max(tmp[k:]))])
#                idx = tmp[k:].index(max(tmp[k:]))
##                print(idx)
#                temp = tmp[k:]
#                temp[0], temp[idx] = temp[idx], temp[0]
#                tmp = tmp[:k] + temp
##                tmp[k], tmp[k:][idx] = tmp[k:][idx], tmp[k]
##                tmp[k:][0], tmp[k:][idx] = tmp[k:][idx], tmp[k:][0]
##                tmp[k], tmp[tmp.index(max(tmp[k:]))] = tmp[tmp.index(max(tmp[k:]))], tmp[k]
##                tmp[k], tmp[k:][tmp[k:].index(max(tmp[k:]))] = tmp[k:][tmp[k:].index(max(tmp[k:]))], tmp[k]
##                print(tmp[k])
##                print(tmp[tmp.index(max(tmp[k:]))])
#                break
#        tmp = [str(x) for x in tmp]
#        tmp = ''.join(tmp)
#        return int(tmp)
        
        nums = list(str(num))
        nums2 = sorted(nums, reverse=True)
        if nums == nums2:
            return num
        else:
            for i in range(len(nums)):
                if nums[i] != nums2[i]:
                    j = -1-nums[::-1].index(nums2[i])
                    nums[i],nums[j] = nums[j],nums[i]
                    break
        return int(''.join(nums))
        
solu = Solution()
num = 2736
#num = 9973
#num = 0
#num = 98368
num = 1993
num2= 9931
print(solu.maximumSwap(num))
